crates/matrix-sdk-crypto/src/error.rs

```diff
@@ -133,6 +133,12 @@ pub enum MegolmError {
     /// The nested value is the sender's current verification level.
     #[error("decryption failed because trust requirement not satisfied: {0}")]
     SenderIdentityNotTrusted(VerificationLevel),
+
+    /// The outer state key could not be verified against the inner encrypted
+    /// state key and type.
+    #[cfg(feature = "experimental-encrypted-state-events")]
+    #[error("decryption failed because the state key failed to validate")]
+    StateKeyVerificationFailed,
```

Comment:
This could be split into a larger variety of cases:

- `MalformedStateKey` - when the outer encrypted event state key is not formatted as `str:str`.
- `MismatchBodyType` - when the decrypted ciphertext is not a state event but the outer event is.
- `EventTypeMismatch` - when the outer event type does not match the inner.
- `StateKeyMismatch` - when the outer state key does not match the inner.

These may not be worth implementing?